apiserver/facades/controller/instancepoller/merge.go

```diff
@@ -153,20 +153,29 @@ func (o *mergeMachineLinkLayerOp) processExistingDevice(dev networkingcommon.Lin
 		return ops, errors.Trace(err)
 	}
 
-	// Warn the user that we will not change a provider ID that is already set.
-	// TODO (manadart 2020-06-09): If this is seen in the wild, we should look
-	// into removing/reassigning provider IDs for devices.
+	// Log a warning if we are changing a provider ID that is already set.
 	providerID := dev.ProviderID()
 	if providerID != "" && providerID != incomingDev.ProviderId {
 		logger.Warningf(
-			"not changing provider ID for device %s from %q to %q",
-			dev.MACAddress(), providerID, incomingDev.ProviderId,
+			"changing provider ID for device %q from %q to %q",
+			dev.Name(), providerID, incomingDev.ProviderId,
 		)
-	} else {
-		ops, err = dev.SetProviderIDOps(incomingDev.ProviderId)
-		if err != nil {
+	}
+
+	ops, err = dev.SetProviderIDOps(incomingDev.ProviderId)
+	if err != nil {
+		if !state.IsProviderIDNotUniqueError(err) {
 			return nil, errors.Trace(err)
 		}
+
+		// If this provider ID is already assigned, log a warning and continue.
+		// If the ID is moving from one device to another for whatever reason,
+		// It will be eventually consistent. E.g. removed from the old device
+		// on this pass and added to the new device on the next.
+		logger.Warningf(
```

Comment:
Just to be safe, can we resort to some sort of ref count for provider IDs while we are constructing the txn ops? It could be something a simple as a `map[ProviderID]int` which we initialize with a value of 1, inc the new id value and dec the old id value. Then, before applying the txn check that the map doesn't have any values > 1.

Reply:
Done. Rather than waiting until the end, we error out the first time we see a provider ID for another device.